### PR TITLE
Runner now accepts a list of option Ids for routing

### PR DESCRIPTION
**Pull request by SamGodwin2**
### What is the context of this PR?
Added the ability to pass in an array of answer values in order to better support checkbox answers.

### How to review 
Tests pass and schemas built using the new functionality work correctly.

### CODE REVIEW

**Comment by ajmaddaford** [comment]
@SamGodwin2 This was added to v3 in https://github.com/ONSdigital/eq-survey-runner/pull/2041

**Comment by SamGodwin2** [comment]
@ajmaddaford Thanks for the heads up. After talking about it though it looks like what came in in that pr doesn't cover all our use cases. So for now we'd like to go with these rules in v2 then switch over when it comes time to move over to v3 later on.

**Comment by SamGodwin2** [comment]
[Spike - Change Runner to accept list of option ID (1 day)](https://trello.com/c/Z7TnIevV/926-spike-change-runner-to-accept-list-of-option-id-1-day)

File: app/assets/js/app/main.js.tmp [hunk 1]
```diff
@@ -0,0 +1,5 @@
+import './modules/inpagelink'
+import './modules/help'
+import './modules/household'
+import './modules/totaliser'
+import './modules/print.js'
```
Comment (pricea1) [inline app/assets/js/app/main.js.tmp]: This file seems to have been added in error

File: app/questionnaire/rules.py [hunk 2]
```diff
@@ -67,14 +67,42 @@ def evaluate_condition(condition, answer_value, match_value):
     comparison_operators = {
         'equals': lambda answer_value, match_value: answer_value == match_value,
         'not equals': lambda answer_value, match_value: answer_value != match_value,
-        'contains': lambda answer_value, match_value: isinstance(answer_value, list) and match_value in answer_value,
-        'not contains': lambda answer_value, match_value: isinstance(answer_value, list) and match_value not in answer_value,
+        'contains':
+            lambda answer_value, match_value: isinstance(answer_value, list)
+            and match_value in answer_value,
+        'all of':
```
Comment (pricea1) [inline app/questionnaire/rules.py]: When launched "All of" fails if other items selected too.

e.g. 
	"condition": "all of",
	"values": ["Box1", "Box2", "Box3"]
passes if Box1, Box2, Box3 selected but
fails if Box1, Box2, Box3, Box4 selected

File: data/en/test_routing_values_array.json [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+                                    "condition": "not all of",
+                                    "values": ["Box1", "Box2", "Box3"]
+                                }]
+                            }
+                        }, {
+                            "goto": {
+                                "block": "block-six",
+                                "when": [{
+                                    "id": "answer-four",
+                                    "condition": "not any of",
```
Comment (pricea1) [inline data/en/test_routing_values_array.json]: Is this being tested? Is this rule only invoked if previous "not all of" rule is false i.e. if Box1, Box2, Box3 all selected.

**Review by pricea1 — CHANGES_REQUESTED** [review]
A couple of logic queries otherwise looks good.

File: data/en/test_routing_values_array.json (hunk 3 above)
Comment (SamGodwin2) [reply data/en/test_routing_values_array.json]: Good spot, its being testing on block 6 I forgot to delete it from here when I moved it.

**Comment by bitdivision** [comment]
I don't feel like I can request changes here, but it seems like this is going to be implementing features in v2 and author which we know will be removed when v3 is merged.

I think it would be better to do any translation required within publisher and use the v3 rules for now.

**Comment by SamGodwin2** [comment]
@bitdivision You do make fair point but while I _think_ it is technically possible to do the translation in publisher the amount of work required to support the range of routing possibilities would be far, far larger. Isn't the routing on the shortlist of things to be changed in v3 anyway so even the ones in there now may not be valid when it comes to merging them both.

**Comment by samiwel** [comment]
It will be a some time before we are able to fully support all v3 features so for now I suggest we strive for consistency in our naming of the comparators (i.e. "contains all", "contains any") and add the inverse comparators not available in v3 (i.e. "not contains all", "not contains any") and implement on the v2 branch.

The additional comparators may need to be implemented on v3 at a later date. But doing it in this order will allow us to deliver on requirements for complex routing conditions on upcoming business surveys and hopefully make the migration to get routing working on v3 less painful.

File: app/questionnaire/rules.py [hunk 4]
```diff
@@ -67,14 +67,42 @@ def evaluate_condition(condition, answer_value, match_value):
...
         'equals': lambda answer_value, match_value: answer_value == match_value,
         'not equals': lambda answer_value, match_value: answer_value != match_value,
-        'contains': lambda answer_value, match_value: isinstance(answer_value, list) and match_value in answer_value,
-        'not contains': lambda answer_value, match_value: isinstance(answer_value, list) and match_value not in answer_value,
+        'contains':
+            lambda answer_value, match_value: isinstance(answer_value, list)
+            and match_value in answer_value,
+        'contains all':
+            lambda answer_value, match_value: isinstance(answer_value, list)
+            and isinstance(match_value, list)
```
Comment (bitdivision) [inline app/questionnaire/rules.py:75]: Should types be checked in here? This seems like a check that  should be done within validator.

File: tests/app/questionnaire/test_rules.py [hunk 5]
```diff
@@ -175,6 +176,40 @@ def test_evaluate_rule_less_than_with_number(self):
         self.assertFalse(evaluate_rule(when, 5))
         self.assertFalse(evaluate_rule(when, 7))
 
+    def test_evaluate_rule_checkbox_contains_all(self):
```
Comment (bitdivision) [inline tests/app/questionnaire/test_rules.py:179]: pytest parameterize is a useful tool on tests like these: https://docs.pytest.org/en/latest/parametrize.html

**Review by pricea1 — APPROVED** [review]
LGTM

File: tests/app/questionnaire/test_rules.py (hunk 5 above)
Comment (SamGodwin2) [reply tests/app/questionnaire/test_rules.py:179]: Would be a nice simplification but unfortunately didn't have any luck in getting it set up and working so decided to leave as is for now.

File: app/questionnaire/rules.py (hunk 4 above)
Comment (SamGodwin2) [reply app/questionnaire/rules.py:75]: They were being checked in here in the existing implementation so I mostly just decided to extend that, also feels like checking them here is simpler that in validator?

**Review by samiwel — APPROVED** [review]
I think the build needs kicking off again, but otherwise LGTM.

**Pull request merged**